libcudacxx/include/cuda/__memory_resource/cuda_memory_pool.h

```diff
@@ -0,0 +1,431 @@
+
+    // allocate the requested initial size to pprime the pool
+    if (__initial_pool_size != 0)
+    {
+      void* __ptr{nullptr};
+      _CCCL_TRY_CUDA_API(
+        ::cudaMallocAsync,
+        "cuda_memory_pool failed allocate the initial pool size",
+        &__ptr,
+        __initial_pool_size,
+        ::cudaStream_t{0});
+
+      _CCCL_ASSERT_CUDA_API(
+        ::cudaFreeAsync, "cuda_memory_pool failed to free the initial pool allocation", __ptr, ::cudaStream_t{0});
+    }
+    return __cuda_pool_handle;
+  }
+
+public:
+  /**
+   * @brief Constructs a cuda_memory_pool with the optionally specified initial pool size
+   * and release threshold.
+   *
+   * If the pool size grows beyond the release threshold, unused memory held by the pool will be
+   * released at the next synchronization event.
+   *
+   * @throws ::cuda::cuda_error if the CUDA version does not support `cudaMallocAsync`
+   *
+   * @param initial_pool_size Optional initial size in bytes of the pool. If no value is provided,
+   * initial pool size is half of the available GPU memory.
+   * @param release_threshold Optional release threshold size in bytes of the pool. If no value is
+   * provided, the release threshold is set to the total amount of memory on the current device.
+   */
+  cuda_memory_pool(
+    const size_t initial_pool_size                             = 0,
+    const size_t release_threshold                             = 0,
+    const cudaMemAllocationHandleType __allocation_handle_type = cudaMemAllocationHandleType::cudaMemHandleTypeNone)
+      : __pool_handle_(__create_cuda_mempool(initial_pool_size, release_threshold, __allocation_handle_type))
+  {}
+
+  cuda_memory_pool(cuda_memory_pool const&)            = delete;
+  cuda_memory_pool(cuda_memory_pool&&)                 = delete;
+  cuda_memory_pool& operator=(cuda_memory_pool const&) = delete;
+  cuda_memory_pool& operator=(cuda_memory_pool&&)      = delete;
+
+  /**
+   * @brief Destroys the `cuda_memory_pool` by releasing the internal cudaMemPool_t.
+   */
+  ~cuda_memory_pool() noexcept
+  {
+    _CCCL_ASSERT_CUDA_API(::cudaMemPoolDestroy, "~cuda_memory_pool() failed to destroy pool", __pool_handle_);
+  }
+
+  /**
+   * @brief Allocate device memory of size at least \p __bytes via cudaMallocAsync.
+   * @param __bytes The size in bytes of the allocation.
+   * @param __alignment The requested alignment of the allocation.
+   * @throws cuda::std::bad_alloc in case of invalid alignment or cuda::cuda_error of the returned error code.
+   * @return Pointer to the newly allocated memory
+   */
+  _CCCL_NODISCARD void* allocate(const size_t __bytes, const size_t __alignment = default_cuda_malloc_alignment) const
+  {
+    // We need to ensure that the provided alignment matches the minimal provided alignment
+    if (!__is_valid_alignment(__alignment))
+    {
+      _CUDA_VSTD_NOVERSION::__throw_bad_alloc();
+    }
+
+    void* __ptr{nullptr};
+    _CCCL_TRY_CUDA_API(
+      ::cudaMallocAsync,
```

Comment:
Shouldn't `cuda_memory_pool::allocate/allocate_async` use `cudaMallocFromPoolAsync? Otherwise the pool created by this class isn't being used for allocation.

Reply:
From the CUDA runtime API:
>__host__​[cudaError_t](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1gf599e5b8b829ce7db0f5216928f6ecb6) cudaMallocAsync ( void** ptr, size_t size, [cudaMemPool_t](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1gd7121b19b2347b777c07223c6ff4cdad) memPool, [cudaStream_t](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1ge15d9c8b7a240312b533d6122558085a) stream )
Allocate from a pool.
Description
This is an alternate spelling for cudaMallocFromPoolAsync made available through operator overloading.

Reply:
We could make it more explicit